src/analyzer.ts

```diff
@@ -8,762 +8,139 @@
-   * of `href` will be ignored.
-   */
-  filter?: (path:string)=> boolean;
-}
-
-/**
- * An Error extended with location metadata.
- */
-interface LocError extends Error{
-  location: {line: number; column: number};
-  ownerDocument: string;
+import * as path from 'path';
+import * as urlLib from 'url';
+
+import {CssParser} from './css/css-parser';
+import {EntityFinder} from './entity/entity-finder';
+import {findEntities} from './entity/find-entities';
+import {HtmlDocument} from './html/html-document';
+import {HtmlImportFinder} from './html/html-import-finder';
+import {HtmlParser} from './html/html-parser';
+import {HtmlScriptFinder} from './html/html-script-finder';
+import {HtmlStyleFinder} from './html/html-style-finder';
+import {ElementFinder} from './javascript/javascript-element-finder';
+import {JavaScriptParser} from './javascript/javascript-parser';
+import {Document} from './parser/document';
+import {Parser} from './parser/parser';
+import {
+  Descriptor,
+  DocumentDescriptor,
+  ImportDescriptor,
+} from './ast/ast';
+import {UrlLoader} from './url-loader/url-loader';
+
+export interface AnalyzerInit {
+  urlLoader?: UrlLoader;
+  parsers?: Map<string, Parser<any>>;
+  entityFinders?: Map<string, EntityFinder<any, any, any>[]>;
 }
 
 /**
  * A database of Polymer metadata defined in HTML
  */
 export class Analyzer {
-  loader: FileLoader;
-  /**
-   * A list of all elements the `Analyzer` has metadata for.
-   */
-  elements: ElementDescriptor[] = [];
-  /**
-   * A view into `elements`, keyed by tag name.
-   */
-  elementsByTagName: {[tagName: string]: ElementDescriptor} = {};
-
-  /**
-   * A list of API features added to `Polymer.Base` encountered by the
-   * analyzer.
-   */
-  features: FeatureDescriptor[] = [];
-
-  /**
-   * The behaviors collected by the analysis pass.
-   */
-  behaviors: BehaviorDescriptor[] = [];
-  /**
-   * The behaviors collected by the analysis pass by name.
-   */
-  behaviorsByName: {[name:string]: BehaviorDescriptor} = {};
-
-  /**
-   * A map, keyed by absolute path, of Document metadata.
-   */
-  html: {[path: string]: AnalyzedDocument} = {};
-
-  /**
-   * A map, keyed by path, of HTML document ASTs.
-   */
-  parsedDocuments: {[path:string]: dom5.Node} = {};
-
-  /**
-   * A map, keyed by path, of JS script ASTs.
-   *
-   * If the path is an HTML file with multiple scripts,
-   * the entry will be an array of scripts.
-   */
-  parsedScripts: {[path:string]: ParsedJS[]} = {};
-
 
-  /**
-   * A map, keyed by path, of document content.
-   */
-  _content: {[path:string]: string} = {};
-
-  /**
-   * @param  {boolean} attachAST  If true, attach a parse5 compliant AST
-   * @param  {FileLoader=} loader An optional `FileLoader` used to load external
-   *                              resources
-   */
-  constructor(attachAST:boolean, loader:FileLoader) {
-    this.loader = loader;
+  private _parsers: Map<string, Parser<any>> = new Map<string, Parser<any>>([
```

Comment:
There's not a common interface between a CSS document node, a JS document node, an HTML document node, produced by three different parsers.